### PR TITLE
fix capi package

### LINKED ISSUE #7181: The size of libmkldnn.so.0 is 0B in capi
Users encounter that lack of libmkldnn.so.0, and they `mv libmkldnn.so libmkldnn.so.0` currently to solve this problem. The reason is that libmkldnn.so.0 is a soft connection.
image
The download address of paddle.tgz is [here](https://guest@paddleci.ngrok.io/repository/download/Manylinux1_CpuAvxCp27cp27mu/.lastSuccessful/paddle.tgz).

### LINKED ISSUE #7254: tar failed when generate capi package
When I use `paddle/scripts/docker/build.sh`  to generate c-api package, it crashed at `tar`.

>  \+ cd /paddle/build/capi_output/usr/local
    + ls
    + egrep -v '^Found.*item$'
    + xargs tar /paddle/build/paddle.tgz
tar: invalid option -- '/'

This error is also encountered [here](https://paddleci.ngrok.io/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=22332&_focus=14704).

The reason that #7237 passed the PR_CI is that `WITH_C_API` was disabled, so it did not actually run into function `gen_capi_package`. 

As teamcity build log  9644 for #7237 shows:

> \+ gen_capi_package
[07:19:56]W:     [Step 1/1] + [[ '' == \O\N ]]
[07:19:56]W:     [Step 1/1] + [[ OFF == \O\N ]]
[07:19:56]W:     [Step 1/1] + printf 'If you need to install PaddlePaddle in de
velop docker image,'
[07:19:56]W:     [Step 1/1] + printf 'please make install or pip install build/python/dist/*.whl.\n'
[07:19:56] :     [Step 1/1] If you need to install PaddlePaddle in develop docker image,please make install or pip install build/python/dist/*.whl.
[07:19:56] :     [Step 1/1]

## how to reproduce this issue:

use below script:
``` shell
#!/bin/bash
set -xe
export PADDLE_DEV_NAME="paddlepaddle/paddle:latest-dev"
docker run -i --rm -v $PWD:/paddle ${PADDLE_DEV_NAME} \
  rm -rf /paddle/build
docker run  -i --rm -v $PWD:/paddle \
  -e "WITH_PYTHON=OFF" \
  -e "WITH_SWIG_PY=OFF" \
  -e "WITH_PYTHON=OFF" \
  -e "WITH_C_API=ON" \
  -e "WITH_GPU=OFF" \
  -e "WITH_AVX=ON" \
  -e "WITH_MKL=ON" \
  -e "WITH_STYLE_CHECK=OFF" \
  ${PADDLE_DEV_NAME} \
  bash -x /paddle/paddle/scripts/docker/build.sh
```

### PR DESCRIPTION
**Pull request by tensor-tang**
fix #7254 and fix #7181

### CODE REVIEW

**Comment by tensor-tang** [comment]
You can directly run the below script to check the result.

```bash
#!/bin/bash
set -xe
export PADDLE_DEV_NAME="paddlepaddle/paddle:latest-dev"
docker run -i --rm -v $PWD:/paddle ${PADDLE_DEV_NAME} \
  rm -rf /paddle/build
docker run  -i --rm -v $PWD:/paddle \
  -e "WITH_PYTHON=OFF" \
  -e "WITH_SWIG_PY=OFF" \
  -e "WITH_PYTHON=OFF" \
  -e "WITH_C_API=ON" \
  -e "WITH_GPU=OFF" \
  -e "WITH_AVX=ON" \
  -e "WITH_MKL=ON" \
  -e "WITH_STYLE_CHECK=OFF" \
  ${PADDLE_DEV_NAME} \
  bash -x /paddle/paddle/scripts/docker/build.sh

```

The paddle.tgz would contain: 
> bin  include  lib  opt  third_party

And the in `lib`:
> lib/libmklml_intel.so
lib/libiomp5.so
lib/libmkldnn.so.0
lib/libpaddle_capi_whole.a
lib/libpaddle_capi_engine.a
lib/libpaddle_capi_layers.a
lib/libpaddle_capi_shared.so

File: cmake/external/mkldnn.cmake [hunk 1]
```diff
@@ -77,3 +77,16 @@ FILE(WRITE ${dummyfile} "const char * dummy = \"${dummyfile}\";")
...
+# copy the real so.0 lib to install dir
+# it can be directly contained in wheel or capi
+SET(MKLDNN_SHARED_LIB ${MKLDNN_INSTALL_DIR}/libmkldnn.so.0)
+ADD_CUSTOM_COMMAND(OUTPUT ${MKLDNN_SHARED_LIB}
+    COMMAND cp ${MKLDNN_LIB} ${MKLDNN_SHARED_LIB}
+    DEPENDS mkldnn)
+ADD_CUSTOM_TARGET(mkldnn_shared_lib ALL DEPENDS ${MKLDNN_SHARED_LIB})
+
+IF(WITH_C_API)
+  INSTALL(FILES ${MKLDNN_SHARED_LIB} DESTINATION lib)
```
Comment (Xreki) [inline cmake/external/mkldnn.cmake]: Not stall to `lib`, but `third_party/mkldnn/lib`. Please refer to https://github.com/PaddlePaddle/Paddle/blob/develop/cmake/external/glog.cmake#L71

File: cmake/external/mklml.cmake [hunk 2]
```diff
@@ -66,3 +66,7 @@ ADD_LIBRARY(mklml SHARED IMPORTED GLOBAL)
 SET_PROPERTY(TARGET mklml PROPERTY IMPORTED_LOCATION ${MKLML_LIB})
 ADD_DEPENDENCIES(mklml ${MKLML_PROJECT})
 LIST(APPEND external_project_dependencies mklml)
+
+IF(WITH_C_API)
+  INSTALL(FILES ${MKLML_LIB} ${MKLML_IOMP_LIB} DESTINATION lib)
```
Comment (Xreki) [inline cmake/external/mklml.cmake]: Not install to `lib`, but `third_party/mklml/lib`.Please refer to https://github.com/PaddlePaddle/Paddle/blob/develop/cmake/external/glog.cmake#L71

File: cmake/external/mkldnn.cmake (hunk 1 above)
Comment (tensor-tang) [reply cmake/external/mkldnn.cmake]: Thanks @Xreki , The reason installed to `lib` is that the script [here ](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/scripts/docker/build.sh#L206) installed to `lib` directly.
This function is trying to replace the command.

And I think the *.so should be contained into `/usr/local/lib`, maybe @Yancey1989 has considered this before?
Anyway I am OK installing to third-party, but then user should set the path to LD manually.

Same as below.

Comment (luotao1) [reply cmake/external/mkldnn.cmake]: As glog etc is a static library, but mkldnn etc is a shared library, thus, if *.so are contained into `thirdparty`, users should `export LD_LIBRARY_PATH` which leads an extra step for users. So we can contain *.so into `/usr/loacl/lib` at this PR.

Comment (Yancey1989) [reply cmake/external/mkldnn.cmake]: The same with @luotao1 , I think `*.so` contained into `/usr/local/lib` is an easy way for the user.

**Pull request merged**